Quests improvements, UI improvements, more game play possibilities...

eauchat opened the pull request:
Well this PR is a refactoring of the way quests are organized and contains quite a lot of content.
Here is an unordered non-exhaustive list of improvements:
- much more game possibilities with less quests
- more context to quest maps, so that regions/cities have context around them making the game experience more instructive
- autofocusing of the region of the map that will be played
- various small ui improvements and possibility to add more info on regions (again for more context on regions/countries/cities found)

Instead of relying on a bunch of different sources for maps, I made use of [natural earth](https://www.naturalearthdata.com/) maps available and very rich in information, and tweaked them to fit the needs of geoquest. So apart from the world countries quest which still uses the same map, other topojson maps are automatically built from one or more geojson files, elements, tags, regions, translations are also automatically derived from the geojson file. I didn't work on it so far, but the richness of info available in the geojson file could make translation of countries/cities/regions names automatic. I think it would be very little work to implement (if I make time for it I'll try to implement it).

I made a slight opinionated change to stop using fonts from google and instead host them directly, this adds about 1MB of storage space, which is more than compensated by saved space in quests maps, and is an improvement for the privacy of users of the site. For me it's a bit important to care for users privacy as part of free software so I hope you're ok for keeping that change :)

I'm still planning on adding more maps, for regions of each continent, for lakes and rivers, for mountains, but I don't know when, depends on my availability to work on it.

Hope you appreciate the contribution :)

woutdp commented:
Oh wow safe to say I need some time to digest this. Thank you for the PR!
Are you still working on it?

woutdp commented:
Ok thank you!

woutdp commented:
The new maps cause the build to exceed 25mb, which is the limit for cloudflare pages. So I need to move some stuff to an R2 bucket to make it work. So we'll be on the previous version for a little bit until I figure that one out

eauchat commented:
Hey hello @woutdp,
Sorry for the delay in the answer. Thanks for merging, I didn't test the result yet but I'll do it when I find time.
About the 25mb limit, two eventual solutions:
 - you could simply filter out all `.geojson` files for what's to be on cloudflare pages, they are necessary to keep in the repo because without them maps are pretty much unreadable and unmodifiable for developping purposes, but the app actually only uses the generated topojson files (the `map.json` in each quest dir), not having `.geojson` files should result in a ~15MB repo at the moment which should work fine
 - or eventually if you want some help on hosting I've got a server running on which it could probably run, let me know if that can be of any use

woutdp commented:
I'm not sure the first point would work, I think it already removes any files that are not used in the app.

I'd be up for using your server if I can get some kind of control of the deployment. I don't have an issue if you re host it though, just anything that goes under geoquest.gg I want to be able to control.

eauchat commented:
Ok, I'll think of it for the server. I have to understand better the security implications for other services. So it'll take me a bit to figure out.
In the meantime maybe it's worth trying the first option. I don't really see any way cloudflare could know which file is used and which not, and in which context for the app. Without those files the repo is surely less than 25MB. With the whole git history it was more than 25MB even before, and with node_modules even more.
I've never used cloudflare, but since it worked before, I guess it counts the repo files without git history and node modules. If it needs a git repository url, I guess one could simply create a branch without the geojson files and make cloudflare fetch that one.
Anyway it's just some random thoughts. See what's best for you.

Also I'll try to reflect on a better color scheme than the one I had proposed. I'm pretty sure we can find something that has enough levels for clarity and that you like. If you have criteria let me know them and I'll try to make some suggestions of color schemes. Or like @SpirioSpirio proposed, I can also try to design an interface to make the theme choosable.

eauchat commented:
Hey hello @woutdp,

In this PR, I had modified the code of how the map is actually displayed, but in reality this introduces a bug for maps of smaller territories where stuff become unclickable. I had made that change because there was some strange behaviour with your previous code where geometries would move after being found and I could not figure out why. Also I didn't manage to implement zooming around the territory of the game with that code at first.
Long story short, I reverted to your code and managed to solve the bug about the geometries moving around when clicked and managed to implement focusing geometries properly as well.
I can make a PR with those changes in the coming days I think.

Also today I've taken a bit of time to review the changes you made to this PR (and maybe others afterwards). Overall I think the UI changes and simplifications you made are nice. And sorry to have forgotten to remove now useless translation files.
But I'm a bit confused about some changes you made to the maps of europe and world cities:
- Sark and Guernsey are not the same thing, [Sark](https://en.wikipedia.org/wiki/Sark) is a part of the [Bailiwick of Guernsey](https://en.wikipedia.org/wiki/Bailiwick_of_Guernsey), so the map is actually mistaken if it shows Sark as this group of islands, calling them Guernsey is actually also not exact (since it's only the bigger island of them which is called this way), but usually people don't say "Bailiwick of Guernsey" and simply Guernsey for simplicity, I believe, but if we want the accurate name I think it would be "Bailiwick of Guernsey".
- I chose to avoid having multiple capital cities for South Africa and Bolivia, even if it's accurate, because if we enter this little game of [multiple capitals](https://en.wikipedia.org/wiki/List_of_national_capitals), in reality it has no ends, and [half of the world countries have multiple capitals](https://en.wikipedia.org/wiki/List_of_countries_with_multiple_capitals). We'd need to add "The hague" in the netherlands, which is the real capital and not Amsterdam, and next would be the turn of Eswatini, Malaysia, Sri Lanka, South Korea, and quite a few little pacific islands as well... To me Bolivia and South Africa make curiosity rise about how capital cities are decided, so it's fun to have multiple capitals, but I'm not so sure it's fair to other countries.
- Also the info I wrote in the `.source` files were actually useful to me for being able to reproduce the map, they are not perfect, but I still tried to give as precise a guideline of what I did, it's also useful sometimes in case for whatever reason it's necessary to redo those steps.

Also may I ask why you removed the lines about which quest the achievements belong to? To me it added some clarity, but don't know, just wondering why you prefer it without.

Thanks for the improvements you made. I can make a PR in the coming days with some little fixes, you can see then if you like them maybe.
